File: iOS/AgoraEntScenarios/Scenes/ShowTo1v1/ShowTo1v1/Classes/Service/ShowTo1v1ServiceProtocol.swift
//
//  ShowTo1v1ServiceProtocol.swift
//  ShowTo1v1
//
//  Created by wushengtao on 2023/7/27.
//

import Foundation


@objc enum ShowTo1v1ServiceNetworkStatus: Int {
    case connecting = 0 // 连接中
    case open // 已打开
    case fail // 失败
    case closed // 已关闭
}

protocol ShowTo1v1ServiceListenerProtocol: NSObjectProtocol {
    
    /// 网络状况变化
    /// - Parameter status: <#status description#>
    func onNetworkStatusChanged(status: ShowTo1v1ServiceNetworkStatus)
    
    /// 用户变化
    /// - Parameter userList: <#userList description#>
    func onUserListDidChanged(userList: [ShowTo1v1UserInfo])
    
    
    /// <#Description#>
    /// - Parameter roomInfo: <#roomInfo description#>
    func onRoomDidDestroy(roomInfo: ShowTo1v1RoomInfo)
}

protocol ShowTo1v1ServiceProtocol: NSObjectProtocol {
    
    /// 创建房间
    /// - Parameters:
    ///   - roomName: <#roomName description#>
    ///   - completion: <#completion description#>
    func createRoom(roomName: String, completion: @escaping (ShowTo1v1RoomInfo?, Error?) -> Void)
    
    /// 加入一个房间
    /// - Parameters:
    ///   - roomInfo: <#roomInfo description#>
    ///   - completion: <#completion description#>
    func joinRoom(roomInfo:ShowTo1v1RoomInfo, completion: @escaping (Error?) -> Void)
    
    /// 离开房间
    /// - Parameters:
    ///   - roomInfo: <#roomInfo description#>
    ///   - completion: <#completion description#>
    func leaveRoom(roomInfo:ShowTo1v1RoomInfo, completion: @escaping (Error?) -> Void)
    
    /// 获取房间列表
    /// - Parameter completion: 完成回调
    func getRoomList(completion: @escaping ([ShowTo1v1RoomInfo]) -> Void)
    
    /// 订阅回调变化
    /// - Parameter listener: <#listener description#>
    func subscribeListener(listener: ShowTo1v1ServiceListenerProtocol?)
    
}

File: Android/scenes/showTo1v1/src/main/java/io/agora/scene/showTo1v1/service/ShowTo1v1ServiceProtocol.kt
package io.agora.scene.showTo1v1.service

enum class ShowTo1v1ServiceNetworkStatus {
    Connecting, // Connecting
    Open,       // Opened
    Fail,       // Failed
    Closed,     // Closed
}

interface ShowTo1v1ServiceListenerProtocol {

    // Network status change callback
    fun onNetworkStatusChanged(status: ShowTo1v1ServiceNetworkStatus)

    // User list change
    fun onUserListDidChanged(userNum: Int)

    // Room destroyed
    fun onRoomDidDestroy(roomId: String)

    // Room time expired
    fun onRoomTimeUp()
}

interface ShowTo1v1ServiceProtocol {

    // Create room
    fun createRoom(roomName: String, completion: (error: Exception?, roomInfo: ShowTo1v1RoomInfo?) -> Unit)

    // Join room
    fun joinRoom(roomInfo: ShowTo1v1RoomInfo, completion: (error: Exception?) -> Unit)

    // Leave room
    fun leaveRoom(roomInfo: ShowTo1v1RoomInfo, completion: (error: Exception?) -> Unit)

    // Get room list
    fun getRoomList(completion: (error: Exception?, roomList: List<ShowTo1v1RoomInfo>) -> Unit)

    // Subscribe to listener
    fun subscribeListener(listener: ShowTo1v1ServiceListenerProtocol)

    // Release resources
    fun reset()
}